mesonbuild/interpreterbase/interpreterbase.py

```diff
@@ -123,7 +123,19 @@ def sanity_check_ast(self) -> None:
             raise InvalidCode('No statements in code.')
         first = self.ast.lines[0]
         if not isinstance(first, mparser.FunctionNode) or first.func_name != 'project':
-            raise InvalidCode('First statement must be a call to project')
+            p = pathlib.Path(self.source_root).resolve()
+            found = p
+            for parent in p.parents:
```

Comment:
This can cause a loop.

Reply:
What is the loop? pathlib returns an iterator over the parent elements as pathlib components, terminating in the filesystem anchor (`/` on Linux, `C:\\` or some other drive on Windows).

It doesn't even resolve symlinks or anything. In fact, it doesn't actually resolve their on disk locations.  It just treats everything as a PurePath string, you get the same (successful) results for a path location that doesn't exist on disk.

Reply:
Ah, I didn't know that. Fair enough.